src/main/kotlin/org/rust/ide/annotator/RsCargoCheckAnnotator.kt

```diff
@@ -147,18 +147,23 @@ private fun filterMessage(file: PsiFile, document: Document, message: RustcMessa
     val spanFilePath = PathUtil.toSystemIndependentName(span.file_name)
     if (!file.virtualFile.path.endsWith(spanFilePath)) return null
 
+    @Suppress("NAME_SHADOWING")
     fun toOffset(line: Int, column: Int): Int? {
+        val line = line - 1
+        val column = column - 1
         if (line >= document.lineCount) return null
-        val result = document.getLineStartOffset(line) + column
-        if (result > document.textLength) return null
-        return result
+        return (document.getLineStartOffset(line) + column)
+            .takeIf { it <= document.textLength }
     }
 
     // The compiler message lines and columns are 1 based while intellij idea are 0 based
-    var textRange = TextRange(
-        toOffset(span.line_start - 1, span.column_start - 1) ?: return null,
-        toOffset(span.line_end - 1, span.column_end - 1) ?: return null
-    )
+    val startOffset = toOffset(span.line_start, span.column_start)
+    val endOffset = toOffset(span.line_end, span.column_end)
+    var textRange = if (startOffset != null && endOffset != null && startOffset < endOffset) {
```

Comment:
But it doesn't solve the main problem, it only allows us to avoid invalid range.
The main problem is that we try to apply the result of `cargo check` to incorrect file with the same suffix of path.
I think we should differ project files from libraries files with the same path suffix and don't annotate incorrect files. In #1498 I suggest annotating only project files as a possible solution.